x-pack/plugins/security_solution/server/lib/telemetry/receiver.ts

```diff
@@ -413,36 +413,75 @@ export class TelemetryReceiver implements ITelemetryReceiver {
     return this.esClient.search(query, { meta: true });
   }
 
-  public async fetchDiagnosticAlerts(executeFrom: string, executeTo: string) {
+  public async *fetchDiagnosticAlerts(executeFrom: string, executeTo: string) {
     if (this.esClient === undefined || this.esClient === null) {
       throw Error('elasticsearch client is unavailable: cannot retrieve diagnostic alerts');
     }
 
-    const query = {
-      expand_wildcards: ['open' as const, 'hidden' as const],
-      index: `${DEFAULT_DIAGNOSTIC_INDEX}-*`,
-      ignore_unavailable: true,
-      size: telemetryConfiguration.telemetry_max_buffer_size,
-      body: {
-        query: {
-          range: {
-            'event.ingested': {
-              gte: executeFrom,
-              lt: executeTo,
-            },
+    let pitId = await this.openPointInTime(DEFAULT_DIAGNOSTIC_INDEX);
+    let fetchMore = true;
+    let searchAfter: SortResults | undefined;
+
+    const query: ESSearchRequest = {
+      query: {
+        range: {
+          'event.ingested': {
+            gte: executeFrom,
+            lt: executeTo,
           },
         },
-        sort: [
-          {
-            'event.ingested': {
-              order: 'desc' as const,
-            },
-          },
-        ],
       },
+      track_total_hits: false,
+      sort: [
+        {
+          'event.ingested': {
+            order: 'desc' as const,
+          },
+        },
+      ],
+      pit: { id: pitId },
+      search_after: searchAfter,
+      size: telemetryConfiguration.telemetry_max_buffer_size,
+      expand_wildcards: ['open' as const, 'hidden' as const],
     };
 
-    return this.esClient.search<TelemetryEvent>(query);
+    let response = null;
+    while (fetchMore) {
+      try {
+        response = await this.esClient.search(query);
+        const numOfHits = response?.hits.hits.length;
+
+        if (numOfHits > 0) {
+          const lastHit = response?.hits.hits[numOfHits - 1];
+          searchAfter = lastHit?.sort;
+        } else {
+          fetchMore = false;
+        }
+
+        tlog(this.logger, `Diagnostic alerts to return: ${numOfHits}`);
+        fetchMore = numOfHits > 0 && numOfHits < telemetryConfiguration.telemetry_max_buffer_size;
+      } catch (e) {
+        tlog(this.logger, e);
+        fetchMore = false;
+      }
+
+      if (response == null) {
+        await this.closePointInTime(pitId);
+        return;
+      }
+
+      const alerts = response?.hits.hits.flatMap((h) =>
+        h._source != null ? ([h._source] as TelemetryEvent[]) : []
+      );
+
+      if (response?.pit_id != null) {
+        pitId = response?.pit_id;
+      }
+
+      yield alerts;
```

Comment:
Perhaps instead of returning a `TelemetryEvent[]` generator it would be more intuitive to return a `TelemetryEvent` generator, or do you prefer the caller to manage each "page" independently? I mean, if we extrapolate it to a java/kotlin `Iterator` the current implementation returns `Iterator<List<TelemetryEvent>>` instead of `Iterator<TelemetryEvent>`, WDYT?

Reply:
I guess it is a matter of taste at this point 🤷🏻 
I'm happy to implement your suggestion if you can be more specific about the benefits.

Reply:
>  I'm happy to implement your suggestion if you can be more specific about the benefits.

The function name is `fetchDiagnosticAlerts` but returns kinda "list of lists", my suggestion was to be consistent between the function name and its return type.

Reply:
Ahh, gotcha. Thanks for the clarification. Will rename